openhtf/__init__.py

```diff
@@ -45,6 +46,106 @@ class InvalidTestPhaseError(Exception):
   """Raised when an invalid method is decorated."""
 
 
+def Test(*phases, **metadata):
+  return TestInfo(phases, metadata=metadata)
+
+
+class TestInfo(mutablerecords.Record(
```

Comment:
I'm okay with that separation - I don't really feel too strongly about it either way.  My point is that there are two methods on this class that have side effects beyond getting/setting its own internal state, `OutputTestRecord()` and `Execute()`.  Making `Execute()` a thin wrapper around a global-scope function doesn't change the fact that calling it has side effects.  `GetPlugTypeMap()` is a great example of what constitutes a 'helper function' on a POD type; in fact you could even make it an `@property` and call it just `plug_type_map` (oh, looking at it, it already was, not sure why you changed this).

Making `Test()` just an arg unpacker for a `TestInfo` record forces a functional user-facing interface over an object-oriented one (ie, make the user call `openhtf.ExecuteTest(foo)` instead of `foo.Execute()`, as you described).  I have a preference for the latter, I think it's more succinct and straightforward, which is why the `Test()` class was structured the way it was in the first place.  I'm okay with pulling out the initialization/state tracking into a separate record class, but the user-facing interface should remain object-oriented, and breaking convention of POD types to provide that interface isn't acceptable.

The two solutions I'm okay with are promoting this to a real class, not a record (but that basically just puts you back where it was before), or keeping a `Test` class whose only purpose is to provide a true object-oriented user interface, and it can keep a `TestInfo` to encapsulate all init/state stuff.  This is generally how functional-to-OO adapters work.